### PR TITLE
CMAD - 20 | Show page

### PR DESCRIPTION
**Pull request by anondita**
Sample DSL used for this:

image

### CODE REVIEW

**Comment by anondita** [comment]
image
Example DSL used:

File: app/views/cm_admin/main/index.html.slim [hunk 1]
```diff
@@ -1,9 +1,10 @@
-== render 'cm_admin/main/top_navbar'
-.index-page
-  - if @model.filters.present?
-    .index-page__filters
-      == render partial: 'cm_admin/main/filters', locals: { filters: @model.filters }
-  .index-page__table-container
-    == render 'cm_admin/main/table'
+.index-page.page-container
```
Comment (anbublacky) [inline app/views/cm_admin/main/index.html.slim:1]: Name the top class with cm-index-page, So it does not get overwritten in application

File: app/views/layouts/cm_admin.html.slim [hunk 2]
```diff
@@ -18,4 +18,23 @@ html
...
+                == render 'cm_admin/main/top_navbar'
+                == render 'cm_admin/main/tabs'
+            .show-page__inner.scrollable
+              = yield
+        - elsif defined?(@action) && @action&.layout_type.to_s == 'cm_association_index'
+          .show-page.page-container
+            .show-page__tabs.sticky-container
+              .cm-tabs-bar
+                == render 'cm_admin/main/top_navbar'
+                == render 'cm_admin/main/tabs'
```
Comment (anbublacky) [inline app/views/layouts/cm_admin.html.slim:34]: Can you check if this can be reused? it looks the same

File: lib/cm_admin/model.rb [hunk 3]
```diff
@@ -58,8 +80,10 @@ def actions(only: [], except: [])
       @actions_set = true
     end
 
-    def cm_show(&block)
-      @current_action = CmAdmin::Models::Action.find_by(self, name: 'show')
+    def cm_show(page_title: ,page_description: ,&block)
+      # @current_action = CmAdmin::Models::Action.find_by(self, name: 'show')
+      @page_title = page_title
+      @page_description = page_description
```
Comment (anbublacky) [inline lib/cm_admin/model.rb:86]: I remember writing a helper, if the page_title or description is empty. Can you check  we can remove this.

File: lib/cm_admin/model.rb [hunk 4]
```diff
@@ -105,6 +129,43 @@ def filter_by(params, filter_params={}, sort_params={})
...
+            sql
+          }.join(' AND '),
+          *terms.map { |e| [e] * filter.db_column_name.size }.flatten
+        )
+      end
+      records
+    end
+
+=======
+>>>>>>> 0ed78242394a8de0a9cafdcff5bf658fd94c17af
```
Comment (anbublacky) [inline lib/cm_admin/model.rb:168]: Remove this, See if everything is merged properly

File: lib/cm_admin/model.rb [hunk 5]
```diff
@@ -147,19 +208,46 @@ def page_description(description)
...
+      else
+        if layout_type.present? && layout.nil? && partial.nil?
+          case layout_type
+          when 'cm_association_index'
+            layout = '/cm_admin/main/associated_index'
+            partial = '/cm_admin/main/associated_table'
+          when 'cm_association_show'
+            layout = '/cm_admin/main/associated_show'
+          end
+        end
```
Comment (anbublacky) [inline lib/cm_admin/model.rb:225]: Shouldn't this be inside action when initialized, check based on layout type?

File: lib/cm_admin/models/cm_show_section.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+      def initialize(section_name, &block)
+        @available_section_fields = []
+        @section_name = section_name
+        puts "-- reached here --"
+        instance_eval(&block)
+      end
+
+      def field(field_name, options={})
+        puts "For printing field #{field_name}"
```
Comment (anbublacky) [inline lib/cm_admin/models/cm_show_section.rb:15]: Remove all the puts statement

File: lib/cm_admin/models/filter.rb [hunk 7]
```diff
@@ -28,8 +28,8 @@ def structure_data(db_column_name, filter_type)
 
       # Methods to filter the records based on the filter type.
       class << self
-        def filtered_data(filter_params, model_name, filters)
-          records = model_name.constantize.where(nil)
+        def filtered_data(filter_params, records, filters)
+          records = self.name.constantize.where(nil) unless records
```
Comment (anbublacky) [inline lib/cm_admin/models/filter.rb:32]: Does filter class have a name?

**Review by anbublacky — CHANGES_REQUESTED** [review]

File: lib/cm_admin/model.rb (hunk 3 above)
Comment (anondita) [reply lib/cm_admin/model.rb:86]: No bro this is for assigning page title throughout the show page tabs

File: app/views/layouts/cm_admin.html.slim (hunk 2 above)
Comment (anondita) [reply app/views/layouts/cm_admin.html.slim:34]: It has a few extra elements bro

File: lib/cm_admin/models/filter.rb (hunk 7 above)
Comment (AdityaTiwari2102) [reply lib/cm_admin/models/filter.rb:32]: @anondita - `self.name` will return the class here. We need the model name against which the records can be fetched from the DB.

**Review by AdityaTiwari2102 — CHANGES_REQUESTED** [review]

**Review by anbublacky — APPROVED** [review]

**Pull request merged**